Wireup date picker

alfetopito opened the pull request:
# Summary

PR on top of https://github.com/cowprotocol/cowswap/pull/1922

- Wiring up the picker with state
- Added error msgs for when:
  - `< min date
  - `> max date
  - Unable to parse inserted date
- Deadline will only be updated if valid, even if input validation/picker is bypassed

⚠️ Error msg is in desperate need of styling. Will delegate that to Michel ;)

<img width="481" alt="Screenshot 2023-01-27 at 15 55 52" src="https://user-images.githubusercontent.com/43217/215130390-7d869d7d-cb10-4a64-a094-0fed1900b608.png">

# To Test

1. On limit orders, open the date picker and select `custom`
2. Play around with the custom date
3. Chose a date < 30min from now
4. Chose a date > 6 months from now

elena-zh commented:
Hey @alfetopito , just wondering why the min date can't be less than 30 min? 
Currently, I can place orders with 2 minutes duration (not less)

elena-zh commented:
And a few comments: 
1. I think that it would be great to make the warning date dynamic (as time is going) 
![image](https://user-images.githubusercontent.com/70885163/215135750-084f8224-37f0-4f98-95e7-ae72e13a4885.png)
2. It would be great to see the default time according to a user's time zone :) 
![image](https://user-images.githubusercontent.com/70885163/215137035-73ace312-0b25-40a0-bc00-21c86a6fcaa5.png)
3. It would be great to open the calendar by clicking on the line, not only on the icon 
![image](https://user-images.githubusercontent.com/70885163/215137378-3e890eb5-9d05-4bf9-86bd-0ca0b3c650d4.png)

alfetopito commented:
> Hey @alfetopito , just wondering why the min date can't be less than 30 min? Currently, I can place orders with 2 minutes duration (not less)

Well, I honestly don't know. This was the default already.

I assume that if you want to use less you can use the pre-defined values? 🤷

alfetopito commented:
> And a few comments:
> 
>     1. I think that it would be great to make the warning date dynamic (as time is going)
>        ![image](https://user-images.githubusercontent.com/70885163/215135750-084f8224-37f0-4f98-95e7-ae72e13a4885.png)

That might be problematic as it can cause infinite re-renders: time changes at every moment, so we'd have to be constantly re-rendering.
Of course, I could set it to update every minute or so, but this is cleared everytime the modal is re-opened, so I don't see it as a big deal

>     2. It would be great to see the default time according to a user's time zone :)
>        ![image](https://user-images.githubusercontent.com/70885163/215137035-73ace312-0b25-40a0-bc00-21c86a6fcaa5.png)

Hmmm, interesting. It was supposed to do that already, so much so that the error is in your locale. Will (try to) fix it
 
>     3. It would be great to open the calendar by clicking on the line, not only on the icon
>        ![image](https://user-images.githubusercontent.com/70885163/215137378-3e890eb5-9d05-4bf9-86bd-0ca0b3c650d4.png)

That's a bit on the browser's court. Every browser handles it differently.
@fairlighteth do you think we could do something about this?

>     4. And it would be nice not to display the error on opening. Only after user's input (if it is not valid one)

Will fix it

alfetopito commented:
1. It wasn't working as I explained. Now it'll be updated every time the `custom` modal is opened
2. Not fixed yet!
4. Fixed

elena-zh commented:
> > Hey @alfetopito , just wondering why the min date can't be less than 30 min? Currently, I can place orders with 2 minutes duration (not less)
> 
> Well, I honestly don't know. This was the default already.
> 
> I assume that if you want to use less you can use the pre-defined values? 🤷

I would align it with BE limitations. Currently, when I create a limit order with expiry value < 2 min, it fails (I reported it in #1417 , case 3). The save min expiration time (2 min) we use for Market orders.

elena-zh commented:
Hey @alfetopito , cases 1, 4 work for me!
However, I have found 1 more issue: 'Clear' button is not working ;(
Can we do anything with it or we can't control its behavior?
![not working](https://user-images.githubusercontent.com/70885163/215459099-32677230-e1f9-4d96-a548-4c2789fa196b.png)

elena-zh commented:
> > Hey @alfetopito , just wondering why the min date can't be less than 30 min? Currently, I can place orders with 2 minutes duration (not less)
> 
> Well, I honestly don't know. This was the default already.
> 
> I assume that if you want to use less you can use the pre-defined values? 🤷

Hey @alfetopito , according to the [discussion](https://cowservices.slack.com/archives/C044B50A02F/p1675087132578169), 10 min should be set as min deadline

alfetopito commented:
@elena-zh 

1. Set deadline to 10min
2. Clear button will reset the date to the min
3. And now the timezone thing is respected (as far as I tested).
But do let me know if there's something still not working.

As for the points in https://github.com/cowprotocol/cowswap/issues/1417

2. Up to the browser. Even if a bad time is picked, the validation shouldn't allow it to be saved
3. Should be addressed here
4. This is also up to the browser unfortunately
5. Same as `2`
6. Same as `2`

shoom3301 commented on src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:
```diff
@@ -145,4 +182,67 @@ export function DeadlineSelector(props: DeadlineSelectorProps) {
 }
 
-// write bash cron job to run this script every day at 00:00
+function _limitDateString(date: Date | number): string {
```
I think it makes sense to move all utility functions to separated file, like a `DeadlineSelector/utils.ts`

shoom3301 commented on src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:
```diff
@@ -42,48 +35,86 @@ export interface DeadlineSelectorProps {
+      const newDeadline = new Date(value).getTime()
+
+      if (newDeadline < minDate.getTime()) {
+        setError(`Must be after ${minDate.toLocaleString()} ${Intl.DateTimeFormat().resolvedOptions().timeZone}`)
```
```typescript
const {timeZone} = Intl.DateTimeFormat().resolvedOptions()
```

shoom3301 commented on src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:
```diff
@@ -42,48 +35,86 @@ export interface DeadlineSelectorProps {
+      console.error(`[DeadlineSelector] Failed to parse input value to Date`, value, e)
+      setError(`Failed to parse date and time provided`)
+    }
+  }, [maxDate, minDate, selectCustomDeadline, value])
```
Should we use Transloco here?

shoom3301 commented on src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:
```diff
@@ -122,20 +153,26 @@ export function DeadlineSelector(props: DeadlineSelectorProps) {
               />
             </CustomLabel>
+            {/* TODO: style me!!! */}
+            {error && <div>{error}</div>}
```
Should we use Transloco here?

alfetopito replied:
What's that?

alfetopito replied:
Oh you mean the `<Trans>` component?

elena-zh commented:
As a nitpick, I'd add an extra padding here:
![image](https://user-images.githubusercontent.com/70885163/216047323-237b49e1-e0fc-4329-b425-1e6c2e6c92ea.png)

alfetopito commented:
> Hey @alfetopito , great job!!! Just 2 issues:
> 
>     1. Reset button is not working in iOS (it does not reset the date to min allowed):
>        ![image](https://user-images.githubusercontent.com/70885163/216044671-9e8fbf91-b462-4402-ac97-d0f204e56721.png)

Fixed.
Could not make the date picker respect the min date, though.
The error message will have to do.

> 
>     2. Previously specified date is saved when switch to custom option:
> 
> 
>     * Select min required Custom date
> 
>     * Then change it to any preselected value (5 min), for example
> 
>     * Wait for at least 1 minute
> 
>     * Select Custom option once again: you will see the warning on opening, the previous value is saved.
>       Can we please reset the date/time to the min required in this case?
>       This is important due to in FF you can't change the time, so the only option to avoid warning is to select the next day.

Fixed. 
Now if the previously selected date is older than the current min date, min date will be used when the modal is opened

> 
>     3. One more usability issue with FF is that the warning appears when I select max available date:
>        ![image](https://user-images.githubusercontent.com/70885163/216046395-bef0689c-4521-4702-9162-5a451a14074c.png)
>        But the condition is met, so I as a user don't understand why the warning is displayed.
>        ![image](https://user-images.githubusercontent.com/70885163/216046768-f124cbf0-2687-4326-8e03-266b47541ecb.png)

That's the expected behaviour IMO.
The date is what it says is allowed, you are right. But you also need to take the time into account.
When you adjust the time the warning should go away.

As for the UI fixes, I expect Michel will make it look much better, not just with padding. I'll leave it up to him once this is approved.

elena-zh commented:
> That's the expected behaviour IMO.
> The date is what it says is allowed, you are right. But you also need to take the time into account.
> When you adjust the time the warning should go away.

Yes, but I described the issue for FF (Firefox): you can't change time there. For all the rest browsers it works as you described.

elena-zh commented:
I have created #1964 issue for the UI issues. 
All the rest changes (besides this [note](https://github.com/cowprotocol/cowswap/pull/1932#issuecomment-1412392595)) LGTM!
